### PR TITLE
Disassemble backwards in variable width instruction sets

### PR DESCRIPTION
**Pull request by OBarronCS**
This implements disassembling backwards in variable width instruction architectures to implement #3784.

Prior to this PR, if you hit a breakpoint or did `nearpc random_address`, and if the disassembly system had not already encountered the surrounding sequence of instructions before, then we were unable to display the "previous" instructions behind the instruction pointer. 

This implements a method to  "guess" the valid sequence of instructions that lead to the current instruction, allowing disassembling backwards. Since ISA's like x86 are not self-synchronizing, we do a best-effort guess at determining the true instruction boundaries.

The method is simple: start a certain amount of bytes in the past, and start disassembling towards the current instruction. We will assume that after a certain number of disassembled instructions, that the sequence will "self-align" to the real sequence (which works in practice).

There is one case where this might not be desirable (i.e. changes longstanding pwndbg behavior), which is stepping into a function call. Previously, when you stepped into a call, the first instruction would be the first one displayed. Now, because we can disassemble backwards, it displays the instructions linearly behind the instruction.

This can get confusing, giving we are now mixing emulation (displaying the true sequence of instruction) with sometimes displaying instructions linearly behind the address.

image

image

### CODE REVIEW

**Comment by OBarronCS** [comment]
The test failures are from now disassembling backwards where we previously couldn't. I'm considering adding a setting for the tests like `disable_backwards_heuristic_disassembly" for the tests, rather than changing the 75+ that have changed.

**Comment by k4lizen** [comment]
> The test failures are from now disassembling backwards where we previously couldn't. I'm considering adding a setting for the tests like `disable_backwards_heuristic_disassembly" for the tests, rather than changing the 75+ that have changed.

I think that's fine, though would be nice to then have some tests that cover this behavior specifically.

**Comment by k4lizen** [comment]
I'm thinking if we should only disassemble backwards until we hit a new symbol, or if it makes sense to do it this way where you can see behind the function you call into.

**Comment by k4lizen** [comment]
> This can get confusing, giving we are now mixing emulation (displaying the true sequence of instruction) with sometimes displaying instructions linearly behind the address.

We could color code them (greying them out seems quite intuitive) or set a marker right above the PC that denotes this maybe (something like `┻`? to signal "everything above this is linear").

**Comment by k4lizen** [comment]
btw i didn't check the source yet but do you use function beginnings as a heuristic to help with disassembling backwards

**Comment by OBarronCS** [comment]
It currently doesn't check function boundaries/symbols.

image

**Comment by OBarronCS** [comment]
~~I checked the tests that failed and played around with it more: before merging, I think that in the context disasm view, we should cut off disassembling backwards at function boundaries (dozens of the tests rely on this behavior), and I think it would be most natural (or maybe make this configurable).~~

I think it would be nice to make it configurable if, on for the context disasm view, we want to automatically disassemble backwards beyond the start of the current symbol.

This would rely on the ability to accurately function boundaries even without debugging symbols.

But in the current state, it's fine.

One change that does need to be made is that in the context disasm view, the number of instructions that are disassembled backwards from the heuristics needs to be capped: it's not nice to have half the screen filled with these grayed out instructions when jumping into a function call when `context-disasm-lines` is a large number.

**Comment by k4lizen** [comment]
makes sense

**Comment by OBarronCS** [comment]
I added a cap to the max number of linearly disassembled backwards instructions are displayed in the context view (with a config to change it).

The logic is:
```
 if disasm_backwards_linear_lines == "auto":
      max_backwards_linear_count = min(10, total_lines_on_screen // 3)
  else:
      max_backwards_linear_count = int(disasm_backwards_linear_lines)
```

File: pwndbg/commands/context.py [hunk 1]
```diff
@@ -1335,6 +1335,12 @@ def get_regs(in_regs: list[str | VisitableRegister | None] | None = None) -> lis
     "context-disasm-lines", 10, "number of additional lines to print in the disasm context"
 )
 
+disasm_backwards_linear_lines = pwndbg.config.add_param(
+    "context-disasm-back-linear-lines",
+    "auto",
+    "maximum number of lines to disassemble backwards linearly in memory",
+)
```
Comment (OBarronCS) [inline pwndbg/commands/context.py:1342]: I couldn't find any examples of other config variables that are both integers but could also be special word "auto" for a default behavior

**Comment by OBarronCS** [comment]
image

image

**Comment by OBarronCS** [comment]
There are so many different callsite for the disassembly. The backwards disassembly is breaking tests. A global disable during the tests for the feature is going to be the easiest way to not break all existing tests.

**Comment by OBarronCS** [comment]
Added a test for this feature + some edge cases I came across. This is ready to go. 

To summarize, this PR adds two closely related features:

1. In the context disassembly view, we used to ONLY show instructions that were disassembled as part of the emulated disassembly flow. Now, we can disassemble "backwards" from the PC linearly, to show the linearly previous instructions, which are displayed greyed out to indicate this. See screenshots earlier in this PR

3. A method to disassemble instructions "backwards" in memory. So if you do "nearpc *random_address", or `continue` the program until a breakpoint, it uses a "heuristic" method to disassemble instructions backwards in memory. This is referred to as a "heuristic" in the code, since due to variable length instructions sets there is no perfect way to do it.

The two new configs.

```python
# In the context disasm view, the max amount of how many grayed out "linear" lines disassembled backwards that you can display. "auto" is a nice default, with a max of a 3rd of the total instructions or 10 instructions, whichever is smallest
set context-disasm-back-linear-lines <number|auto> # auto by default
# Disable the heuristic all together
set heuristic-backwards-disasm <on/off>                 # on by default
```

**Review by k4lizen — COMMENTED** [review]
can we get rid of `heuristic-backwards-disasm` and just use `context-disasm-back-linear-lines == 0`?

**Comment by OBarronCS** [comment]
> can we get rid of `heuristic-backwards-disasm` and just use `context-disasm-back-linear-lines == 0`?

The heuristic is for disassembling backwards when we know don't instruction flow ahead of time. This setting enables it in ALL disassembly contexts where we are disassembling around a point.

The `context-disasm-back-linear-lines` setting is specifically is that "for the context disasm view, show this many grayed out linear backwards instructions".

We may want to show 0 lines in the context view, but for normal "nearpc" command invocation, still have it do the backwards disassembly. Having the two settings makes this possible.

File: pwndbg/aglib/disasm/disassembly.py [hunk 2]
```diff
@@ -127,6 +130,14 @@ class InstructionSequenceSavePointer:
     node: InstructionSequenceNode | None
 
 
+# Dict of Address -> previous instruction sequentially in memory
+# Some architectures don't have fixed-sized instructions, so this is used
+# to disassemble backwards linearly in memory for those cases
+linear_backward_address_cache: collections.defaultdict[int, int] = collections.defaultdict(
+    lambda: None
```
Comment (disconnect3d) [inline pwndbg/aglib/disasm/disassembly.py:137]: shouldn't this be: `lambda: 0`? if e map int to int?

Comment (disconnect3d) [inline pwndbg/aglib/disasm/disassembly.py:137]: Or should the type be: `collections.defaultdict[int, int | None]`

Comment (OBarronCS) [reply pwndbg/aglib/disasm/disassembly.py:137]: You're right, it should be type `int | None`. Fixed this.
